### DISCUSSION TITLE
Rename File On Upload? Not Overwriting Other Uploads

### DISCUSSION THREAD

**Discussion by andrewstanton** [payloadcms/payload#1153] [Q&A] [answered]
To prevent uploads from overwriting each other in our media folder, is there a way to generate a unique name for a file upon upload?

> Example:
> Uploading `unknown.jpg` (again) would be renamed to `[random-hash].jpg` on the server

Would this be accomplished by means of a middleware or something I'm missing?

Aka: `handlers` in the Upload Options described here: https://payloadcms.com/docs/upload/overview#enabling-uploads

Thank you for responses / feedback!

**Comment by DanRibbens** [comment]
Uploaded files having the same name as something already in storage will be renamed with a `-n` suffix for uniqueness such that `unknown.jpg` uploaded a 2nd time is saved as `unknown-1.jpg`. This should be happening automatically for you. If that is not the case you can open a bug for this issue.

If you prefer to have a unique hash instead, you'd need to do a bit of work which I can go into some detail if you want.

Let me know what you think!

**Comment by SimonVreman** [comment]
Adding to this, it looks like a [`beforeOperation`](https://payloadcms.com/docs/hooks/collections#beforeoperation) hook on the collection is also suitable for this. A quick and dirty example:
```js
beforeOperation: [async ({ args }) => {
  const files = args.req?.files;
  if (files && files.file && files.file.name) {
    const parts = files.file.name.split('.');
    files.file.name = `${(Math.random() + 1).toString(36).substring(2)}.${parts[parts.length - 1]}`;
  }
}]
```
It just replaces the filename in the request before anything is saved. Note that this does not at all guarantee that the filename is unique, although depending on usage collision are probably rare anyway and in case they do happen we can rely on the built-in functionality of adding the `-n` suffix as Dan pointed out.

Using `handlers` on the upload might be 'better', but I prefer using a hook to stay consistent with the rest of my application.

**Reply by jmikrut** [reply]
+1. This is probably what I'd do. Hell yeah, `beforeOperation`! The use cases for that one are less than the `beforeChange`, etc. but I always knew there were use cases out there!!!

**Reply by andrewjroberts** [reply]
fwiw stumbled on this discussion searching for a way to encode upload filenames that had spaces in them.

good for this purpose too right? 

```ts
hooks: {
    beforeOperation: [
      async ({ args }) => {
        const files = args.req?.files;
        if (files && files.file && files.file.name) {
          files.file.name = encodeURIComponent(files.file.name);
        }
      }
    ]
  },

```

**Comment by christian-reichart** [comment, marked as answer]
I found this thread because I want to sanitize filenames to not include spaces.
Since uploads always have a URL attached, should this maybe be a core functionality?

something like `sanitizeFilename: true`?

Solved it like this if anyone cares:

```javascript
import { CollectionBeforeOperationHook } from 'payload/types'

/**
 * Sanitizes the filename for upload collections to not include spaces or special characters
 */
export const sanitizeFileName: CollectionBeforeOperationHook = async ({
  args,
}) => {
  const files = args.req?.files
  if (
    files !== undefined &&
    files !== null &&
    typeof files === 'object' &&
    files.file !== undefined &&
    files.file !== null &&
    typeof files.file === 'object' &&
    files.file.name !== undefined &&
    files.file.name !== null &&
    typeof files.file.name === 'string'
  ) {
    files.file.name = files.file.name.replace(/[^a-z0-9.]/gi, '_').toLowerCase()
  }
}
```

**Reply by ccssmnn** [reply]
I've tried a `beforeChange`-Hook to manipulate the filename in the document. This changed the document entry, but not the actual file name. Also, the resized image names were still based on the original name.

I now use this hook here to set a nanoid as the filename. Thank you so much!

**Reply by linobino1** [reply]
Using this in Payload `v3.0.0-beta.19`:


```ts
import type { CollectionBeforeOperationHook } from "payload/types";
import slugify from "slugify";

/**
 * Sanitizes the filename for upload collections to not include spaces or special characters
 */
export const sanitizeFileName: CollectionBeforeOperationHook = async ({
  req,
}) => {
  const file = req.file;
  if (typeof file?.name === "string") {
    file.name = slugify(file.name, { strict: true });
  }
}
```

**Comment by wkentdag** [comment]
IMO the [options](https://github.com/parshap/node-sanitize-filename?tab=readme-ov-file#api) passed to `sanitize-filename` (invoked [here](https://github.com/payloadcms/payload/blob/main/packages/payload/src/uploads/getSafeFilename.ts#L10)) should be configurable. For context there's some discussion in discord [here](https://discord.com/channels/967097582721572934/1231361822146170900/1231361822146170900)

**Comment by vojtech-cerveny** [comment]
Anything from suggestions works for me, so I needed to debug it and this works for me. 
Version `3.0.0-beta.58`
```typescript
import { CollectionConfig } from 'payload'

export const Media: CollectionConfig = {
  slug: 'media',
  upload: true,

  fields: [
    {
      name: 'text',
      type: 'text',
    },
  ],
  hooks: {
    beforeOperation: [
      async (req) => {
        if (req.operation === 'create') {
          if (req.req.file)
            req.req.file.name = req.req.file?.name.replace(/[^a-z0-9.]/gi, '').toLowerCase()
        }
      },
    ],
  },
}

```

**Comment by valentinilas** [comment]
An example with UUID

`icon.png` becomes `d02183e2-755d-4518-b290-c4ab7339b9e3.png`

```import type { CollectionConfig } from 'payload'
import { v4 as uuidv4 } from 'uuid';  // Import the UUID function

export const Media: CollectionConfig = {
  slug: 'media',
  access: {
    read: () => true,
  },

  fields: [
    {
      name: 'alt',
      type: 'text',
      required: true,
    },
  ],
  hooks: {
    beforeOperation: [
      async (req) => {
        if (req.operation === 'create') {
          if (req.req.file) {
            // Generate a new UUID for the file name
            const fileExtension = req.req.file.name.split('.').pop(); // Extract file extension
            const newFileName = `${uuidv4()}.${fileExtension}`; // Generate UUID and append the file extension
            req.req.file.name = newFileName; // Assign the new name
          }
        }
      },
    ],
  },
  upload: true,
}
```

**Reply by jeromewtf** [reply]
> An example with UUID
> 
> `icon.png` becomes `d02183e2-755d-4518-b290-c4ab7339b9e3.png`
> 
> ```
> import { v4 as uuidv4 } from 'uuid';  // Import the UUID function
> 
> export const Media: CollectionConfig = {
>   slug: 'media',
>   access: {
>     read: () => true,
>   },
> 
>   fields: [
>     {
>       name: 'alt',
>       type: 'text',
>       required: true,
>     },
>   ],
>   hooks: {
>     beforeOperation: [
>       async (req) => {
>         if (req.operation === 'create') {
>           if (req.req.file) {
>             // Generate a new UUID for the file name
>             const fileExtension = req.req.file.name.split('.').pop(); // Extract file extension
>             const newFileName = `${uuidv4()}.${fileExtension}`; // Generate UUID and append the file extension
>             req.req.file.name = newFileName; // Assign the new name
>           }
>         }
>       },
>     ],
>   },
>   upload: true,
> }
> ```

How to add the possibility to preserve the originalFile in a field? beforeOperation has no access to data :-(

**Reply by giandomenicodisalvatore** [reply]
@jeromewtf  I know I'm late to the party, I did it using **request context** to pass the `originalFilename` to the request data in the `afterOperation` hook. Not sure if it is the right way to do it, **first time user** here

```ts
import crypto from 'crypto'

export const Media: CollectionConfig = {
   // ...
  admin: {
    useAsTitle: 'originalFilename',
  },
   fields: [
   // ...
    {
      name: 'originalFilename',
      type: 'text',
    },
  ],
  hooks: {
    beforeOperation: [
      ({ operation, req }) => {
        if (operation === 'create' && req.file) {
          const [originalFilename, ext] = req.file.name.split('.')
          req.context.originalFilename = `${originalFilename}.${ext}`
          req.file.name = `${crypto.randomUUID()}.${ext}`
        }
      },
    ],
    afterOperation: [
      ({ operation, req }) => {
        if (
          operation === 'create' &&
          req.context?.originalFilename &&
          req.data
        ) {
          req.data.originalFilename = req.context.originalFilename
        }
      },
    ],
  },
}
```

**Reply by SimonVreman** [reply]
@giandomenicodisalvatore `afterOperation` will not persist it; the database operation has already been completed. [`beforeChange`](https://payloadcms.com/docs/hooks/collections#beforechange) would allow you to add the value to the request, make sure to validate it still however. Using context is indeed a good way to solve this cleanly.

**Reply by giandomenicodisalvatore** [reply]
Nice catch @SimonVreman thank you!
